### ISSUE TITLE
Replace all $.ajax and POJOs in model hooks with Ember Data

### ISSUE THREAD

**Issue by michaelrkn** [emberjs/guides#542]
Throughout the guides, sometimes the examples show the `model` hook returning a POJO, sometimes the results of a `$.ajax` call, and sometimes an Ember Data object. The Guides are meant to illustrate the happy path, which in this case is Ember Data.

**Comment by michaelrkn** [comment]
Also, we will document how to use `$.ajax` appropriately in Ember via https://github.com/emberjs/guides/issues/449#issuecomment-127652021.

**Comment by tomdale** [comment]
I think a sidebar explaining what's going on under the hood for the `model()` hook is absolutely needed, but the happy path and examples benefit tremendously in consistency if that hook is always returning a model object.

More important than consistency, though, is that this sets people up for building an app that scales much better as their needs increase, and adds almost no additional complexity.

The gist of it is that, rather than returning a POJO from the model hook, you can push the data into the store first. (Maybe we should consider some syntactic sugar for exactly this case? Maybe it already exists, @igorT @fivetanley @wecc @bmac?) This is what we often to refer to as the "proactive" case—you proactively push data into the store, giving the user a clear place to decorate the data with CPs and additional behavior if they need it. It also means that the answer to "What is the `model` in my template?" is always "An Ember Data model" instead of "who knows?"

For what it's worth, I think @wycats had advocated for making the happy path the "reactive" path (implement an adapter `findRecord()` method) which honestly is not too much additional work, may be better, and can almost certainly be made into a generator that does the right thing by default. So maybe we should try both approaches and see which one strikes the best balance of simplicity and maintainability/growth potential.

**Comment by eccegordo** [comment]
It strikes me that showing Ember Data centric examples for consistency is ok. But I think it is very worthwhile to have a dedicated section of the guides that discusses the "model" layer in more generic terms. I think there is a lot of value in demonstrating that a model hook can return more than just ember data model instances. I would suggest a dedicated section that shows these techniques (ED Model/Store, POJO, $ajax, fixture data, Ember Object, array of strings numbers etc) side by side. I think it is important to make users aware that technically the model hook can return any kind of "data" needed. Synchronous and Asynchronous. Promises and Non-Promises.

And we should be fair in making it clear that there are legitimate alternatives to Ember Data, that serve specific niche use cases.

I know in my experience I have found using vanilla Ember Objects within a Service Object a very powerful pattern.

I think a dedicated section of the guides that contrasts the different ways to model data is very critical to illuminate the subtle distinctions. And @tomdale is right there should be a side bar that explains the intricacies of the model hook.

Think of it this way, not every iOS app uses CoreData. There are other perfectly legitimate alternative approaches.

Ember Data and non ED approaches can live in harmony even within the same app. 

By the way I love ember data, and use it in my apps, so not a criticism. I just find that thinking more expansively and abstractly about the model layer has allowed me to remain flexible and think more creatively on how to use ember to solve interesting use cases.

One person's "happy path" is another person's straight jacket.

**Comment by eccegordo** [comment]
Also, I would suggest for everybody's sanity that you make it crystal clear that the `store` is an Ember Data specific construct. And that when using Ember without ED it is BYOS (bring your own store). That will help people trying to grok local storage, and other esoteric use cases.

**Comment by acorncom** [comment]
> I would suggest a dedicated section that shows these techniques (ED Model/Store, POJO, $ajax, fixture data, Ember Object, array of strings numbers etc) side by side. I think it is important to make users aware that technically the model hook can return any kind of "data" needed. Synchronous and Asynchronous. Promises and Non-Promises.

I would agree with the above.  @tomdale's comments about the "happy path" make total sense (and make other powerful features of Ember Data available).  But explaining what is actually happening with models (even if we say something along the lines of "here be dragons") helped me have a much clearer mental model to understand what was occurring.  Before that, it felt a bit like "magic" :-)

Toran Billup's test driven development talk was the first time I fully understood how flexible the model portion of Ember was, which has helped in prototyping, even when I knew I was headed toward Ember Data ...

**Comment by locks** [comment]
> I would suggest a dedicated section that shows these techniques (ED Model/Store, POJO, $ajax, fixture data, Ember Object, array of strings numbers etc) side by side.

Write a blog post, tell emberwatch and Ember Weekly about it.

> I think it is important to make users aware that technically the model hook can return any kind of "data" needed.

I think we do at `routing/specifying-a-routes-model`. I think it can be done better, but I'd still replace the `$.getJSON`s with ED feeding alternatives, which is part of what OP means.

> And we should be fair in making it clear that there are legitimate alternatives to Ember Data, that serve specific niche use cases.

Not the guides' job.

**Comment by greyhwndz** [comment]
> I would suggest a dedicated section that shows these techniques (ED Model/Store, POJO, $ajax, fixture data, Ember Object, array of strings numbers etc) side by side. I think it is important to make users aware that technically the model hook can return any kind of "data" needed. Synchronous and Asynchronous. Promises and Non-Promises.

I think this may be a good thing for the learning experience of the those who continues or starts to engage in developing ambitious Ember apps. I think part of delivering a framework to the community to create ambitious Ember apps is the delivery of an ambitious Ember guide that will truly serve as a 'guide' on aspects that the framework is trying to make it easy for the developer. I would like to point out that the term 'ambitious' would traverse the aspects of the power and flexibility of the architecture and patterns. It is constantly trying to refine/enhance if not totally innovate on new fronts based on the collective experience, pains & the lessons learned by the developers who use Ember and contribute to the project. Conveying this as part of the 'guides' is I think justified and inevitable lest we come up short on that critical aspect of making it easy for developers to implement their innovative ideas to come up with the ambitious apps we say Ember can be used for.

**Comment by greyhwndz** [comment]
I would just like to clarify my line of thinking and intentions in my comment above because it may appear 'demanding' to some or worse a negative 'critic' of what we have.  Since we have an opportunity to come up with something of an improvement with the guides, I thought I'd give a shot of sharing my line of thought as to why things could be done to improve things. I hope this won't be taken as shooting ideas down or putting what has been brought up by the contributions of the community. My impression of the community is of an innovative one open to ideas and has a culture of respect and the assumption of good intentions. I understand that we still have so many things in the development pipeline and some may see my comments & suggestions above as being more demanding from that angle. I am aware of that but I hope that this should not be taken in that light otherwise seems like my only option is to stop contributing ideas and stop collaborating. I am still hopeful that we could continue a healthy discussion on things like these coz I believe this is one that drives innovative directions. Still I would like to hear feedback if this is going too strong for anybody coz I want to be consider doing something about that.

@locks: Am sorry if I have offended but I have no intention to. Hope I have clarified myself. I have inserted some modifications to my thoughts above.

**Comment by MarkMT** [comment]
I agree that the guides should make it clear that the model() hook is not inherently bound to Ember Data and can return something as simple as a POJO literal. My reason for saying this is less to do with wanting to identify the different ways in which the framework can be used in practice than a belief that it's important for people who are new to the framework to be able to see the key architectural abstractions it's based on. Focusing on 'how-to' and the 'happy path' is great, but it can sometimes obscure that perspective.

The [strawman](https://github.com/MarkMT/guides/tree/concepts) Core Concepts section I wrote for #585 tries to illustrate this ([text](https://github.com/MarkMT/guides/blob/concepts/source/getting-started/core-concepts.md)). The idea of a controller getting data from the route is important architecturally regardless of what the implementation of the hook looks like.

**Comment by locks** [comment]
@greyhwndz like we discussed on Slack, I was not offended, and we're trying to make the guides development as public as we can exactly so people can give their feedback which is valuable for the team.
What I suggested is trying to focus on actionable items (e.g. "I suggest this structure" and not "this structure should be improved") so we can have discussions about it and improve. I don't think anyone will argue that the guides should be continuously improved to be as good as they can be, otherwise we wouldn't have formed a documentation subteam specifically to deal with that. We want ambitious guides, the question is the how and what to improve :)
